src/libsyntax/ext/tt/macro_parser.rs

```diff
@@ -492,20 +480,27 @@ pub fn parse(sess: &ParseSess, rdr: TtReader, ms: &[TokenTree]) -> NamedParseRes
     }
 }
 
-pub fn parse_nt<'a>(p: &mut Parser<'a>, sp: Span, name: &str) -> Nonterminal {
+fn parse_nt<'a>(p: &mut Parser<'a>, sp: Span, name: &str) -> Nonterminal {
     match name {
         "tt" => {
             p.quote_depth += 1; //but in theory, non-quoted tts might be useful
             let mut tt = panictry!(p.parse_token_tree());
             p.quote_depth -= 1;
-            loop {
-                let nt = match tt {
-                    TokenTree::Token(_, token::Interpolated(ref nt)) => nt.clone(),
-                    _ => break,
-                };
-                match *nt {
-                    token::NtTT(ref sub_tt) => tt = sub_tt.clone(),
-                    _ => break,
+            while let TokenTree::Token(sp, token::Interpolated(nt)) = tt {
+                match Rc::try_unwrap(nt) {
+                    Ok(token::NtTT(sub_tt)) => tt = sub_tt,
+                    Ok(nt) => {
+                        tt = TokenTree::Token(sp, token::Interpolated(Rc::new(nt)));
```

Comment:
I believe this recreates the `Rc` that we just unwrapped. Perhaps this could be avoided with something like:
```rust
            while let TokenTree::Token(sp, token::Interpolated(nt)) = tt {
                if let token::NtTT(..) = *nt {
                    match Rc::try_unwrap(nt) {
                        Ok(token::NtTT(sub_tt)) => tt = sub_tt,
                        Ok(_) => unreachable!(),
                        Err(nt_rc) => match *nt_rc {
                            token::NtTT(ref sub_tt) => tt = sub_tt.clone(),
                            _ => unreachable!(),
                        },
                    }
                } else {
                    tt = TokenTree::Token(sp, token::Interpolated(nt.clone()));
                    break
                }
            }
```